src/shogun-core-main/src/main/java/de/terrestris/shoguncore/service/GeoServerInterceptorService.java

```diff
@@ -348,16 +366,38 @@ public Response interceptWmtsRequest(HttpServletRequest request, String serviceI
             throw new InterceptorException("No WMTS request path found!");
         }
         String path = matcher.group(1);
-
-        // get all layers allowed for this user in order to filter out not allowed ones
-        List<Layer> layers = layerService.findAll();
         WmtsLayerDataSource dataSource = null;
-        for (Layer layer : layers) {
-            if (layer.getSource() instanceof WmtsLayerDataSource) {
-                WmtsLayerDataSource source = (WmtsLayerDataSource) layer.getSource();
-                if (source.getId().equals(id)) {
-                    dataSource = source;
-                    break;
+        // get all layers allowed for this user in order to filter out not allowed ones
+        User user = userService.getUserBySession();
+        if (user == null) {
+            // check for anonymous access
+            List<Layer> layers = layerDao.findAll();
+            for (Layer layer : layers) {
+                if (layer.getSource() instanceof WmtsLayerDataSource) {
+                    WmtsLayerDataSource source = (WmtsLayerDataSource) layer.getSource();
+                    if (source.getId().equals(id)) {
+                        Map<UserGroup, PermissionCollection> userGroupPermissions = layer.getGroupPermissions();
+                        if (userGroupPermissions != null) {
+                            for (UserGroup group : userGroupPermissions.keySet()) {
+                                if (group.getName().equalsIgnoreCase("public") &&
+                                    group.getOwner() == null) {
+                                    dataSource = source;
+                                    break;
```

Comment:
This will break out the inner loop only. Please make either use of labels or move this code to a dedicated function that will return the result value (<- preferred).